File: src/structure/java/stack/ArrayStack.java
package structure.java.stack;

import structure.Stack;
import structure.java.ArrayList;

public class ArrayStack<E> implements Stack<E> {
    private ArrayList<E> array;

    public ArrayStack(int capacity){
        array = new ArrayList<>(capacity);
    }

    public ArrayStack() {
        array = new ArrayList<>();
    }

    @Override
    public int getSize() {
        return array.getSize();
    }

    @Override
    public boolean isEmpty() {
        return array.isEmpty();
    }

    @Override
    public void push(E element) {
        array.append(element);
    }

    @Override
    public E pop() {
        return array.pop();
    }

    @Override
    public E peek() {
        return array.getLast();
    }

    public int getCapacity(){
        return array.getCapacity();
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("stack: ");
        stringBuilder.append('[');
        for (int i = 0; i < getSize(); i++) {
            stringBuilder.append(array.get(i));
            if (i != getSize() - 1)
                stringBuilder.append(", ");
        }
        stringBuilder.append("] top");
        return stringBuilder.toString();
    }
}

File: src/structure/kotlin/stack/ArrayStack.kt
package structure.kotlin.stack

import structure.Stack
import structure.java.stack.ArrayStack
import structure.kotlin.ArrayList

class ArrayStack<E> constructor(capacity: Int = 10) : Stack<E> {
    private val array: ArrayList<E> = ArrayList(capacity)
    val capacity: Int
        get() = array.capacity

    override fun getSize(): Int = array.size

    override fun isEmpty(): Boolean = array.isEmpty

    override fun push(element: E) = array.append(element)

    override fun pop(): E = array.pop()

    override fun peek(): E = array.getLast()

    override fun toString(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append("stack: ")
        stringBuilder.append('[')
        for (i in 0 until getSize()) {
            stringBuilder.append(array[i])
            if (i != getSize() - 1)
                stringBuilder.append(", ")
        }
        stringBuilder.append("] top")
        return stringBuilder.toString()
    }
}


// Test
fun main(args: kotlin.Array<String>) {
    val stack = structure.kotlin.stack.ArrayStack<Int>()
    for(i in 1..5){
        stack.push(i)
    }
    println(stack)
    println(stack.pop())
    println(stack)

    val stackJava = ArrayStack<Int>()
    for(i in 1..5){
        stackJava.push(i)
    }
    println(stackJava)
    println(stackJava.pop())
    println(stackJava)
}